### ISSUE TITLE
API / endpoint for snupkg's

### ISSUE THREAD

**Issue by clairernovotny** [NuGet/NuGetGallery#7912]
We need an API / metadata on the package that points to a snupkg, if the package has one.

The scenario is that tooling may need to download all symbols for a NuGet package as part of validation steps, such as ensuring the Source Link information is properly embedded.

Currently the only option is to use the CDN endpoint, and check with a HEAD:  `https://globalcdn.nuget.org/symbol-packages/{lowercase ID}.{lowercase normalized version}.snupkg`

There's no guarantees around this endpoint, however.

Downloading each PDB via the symbol server API is less efficient as each pdb needs to be calculated and downloaded, if published, individually. For validation scenarios, all pdb's would be required anyway.

--

There's another scenario where people find it generally useful to see what's in the snupkg package, downloading and opening with a tool like NuGet Package Explorer.

--

NuGet Package Explorer intends on helping people know if packages are properly Source Linked and thus needs to perform all these steps: https://github.com/NuGetPackageExplorer/NuGetPackageExplorer/issues/904

**Comment by loic-sharma** [comment]
This is a great idea! I would love to have a similar "debuggable" indicator on nuget.org and Visual Studio. I think at the bare minimum we could add a documented [service index entry](https://docs.microsoft.com/en-us/nuget/api/overview#resources-and-schema) for the .snupkg resource.

> Downloading each PDB via the symbol server API is less efficient as each pdb needs to be calculated and downloaded, if published, individually.

Even with the .snupkg API, wouldn't you need to do this regardless? A package may have .DLLs whose symbols are on the Microsoft symbol server but not in its corresponding symbol package.

**Comment by clairernovotny** [comment]
I think it's important to define what dll's are relevant to be validated. For now, in NPE, I'm just checking ones in `lib` and `runtimes`. I'm not validating files in `tools`, etc. Therefore, I expect all the files to be present in the `snupkg` and not need to check any symbol servers.

**Comment by loic-sharma** [comment]
> . I only support snupkg checking if the package came from NuGet.org and I think that's a huge issue and limitation.

Yup that's very fair.

> I think it's important to define what dll's are relevant to be validated. For now, in NPE, I'm just checking ones in `lib` and `runtimes`. I'm not validating files in `tools`, etc. Therefore, I expect all the files to be present in the `snupkg` and not need to check any symbol servers.

Even with that `lib` and `runtimes` folder limitation, I think you would still need to fallback to symbol servers in certain scenarios. For example, [Microsoft.Extensions.Logging.Abstractions](https://www.nuget.org/packages/Microsoft.Extensions.Logging.Abstractions/3.1.3) has published its symbols on the Microsoft symbol server and does not have a corresponding .snupkg on nuget.org. 

![image](https://user-images.githubusercontent.com/737941/77710023-25e29e00-6f8a-11ea-9fe8-882d853aeccc.png)

> I've just released 5.6.18 of NPE

I just played around with the latest version, it works like a charm! :)

**Comment by clairernovotny** [comment]
Good catch re `Microsoft.Extensions.Logging.Abstractions`. I was really hoping to avoid poling symbol servers as that's a whole other game. Given that those are likely to be only Microsoft packages, I'm inclined to not do anything about it either.

**Comment by loic-sharma** [comment]
> I was really hoping to avoid poling symbol servers as that's a whole other game.

I've found this repository invaluable: https://github.com/dotnet/symstore

I took a crack at this problem a while back when I was trying to prototype a similar feature. Maybe it might help! https://github.com/loic-sharma/NuGet.Quality/blob/6fcb1d2e47568a721d1d6fdf62b161c22ab2ad8a/Program.cs#L162-L188

**Comment by clairernovotny** [comment]
@loic-sharma I just published 5.6.23, which checks the Microsoft symbol server for Microsoft files. It fixes this.

**Comment by loic-sharma** [comment]
> I saw your TODO about ensuring symbols match the PE file... any leads on that?

Yes nuget.org does this validation at .snupkg ingestion. The steps are:

1. Get the checksum entries from the DLL (see [this](https://github.com/NuGet/NuGet.Jobs/blob/26c23697fee363d3133171f71e129cda9b5a3707/src/Validation.Symbols/SymbolsValidatorService.cs#L200-L203))
1. Find a PDB that corresponds to the DLL
1. Check that PDB's hash matches at least one checksum entry from the DLL (see [this](https://github.com/NuGet/NuGet.Jobs/blob/26c23697fee363d3133171f71e129cda9b5a3707/src/Validation.Symbols/SymbolsValidatorService.cs#L219-L240)).
    1. For each checksum entry from DLL:
        1. Grab the [algorithm name](https://github.com/dotnet/runtime/blob/0ac5dc9f7a8524dcec83de8f02a738d51b26f5f3/src/libraries/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/DebugDirectory/PdbChecksumDebugDirectoryData.cs#L15) from the checksum entry
        1. Calculate the PDB's content hash using that hash algorithm name (see [this](https://github.com/NuGet/NuGet.Jobs/blob/26c23697fee363d3133171f71e129cda9b5a3707/src/Validation.Symbols/SymbolsValidatorService.cs#L222-L232))
       1. If that PDB content hash matches a checksum entry, pass! (see [this](https://github.com/NuGet/NuGet.Jobs/blob/26c23697fee363d3133171f71e129cda9b5a3707/src/Validation.Symbols/SymbolsValidatorService.cs#L235)) 😀
    1. If no PDB content hash matches the checksum entry, fail! 😠

Please let me know if you have any questions

**Comment by TheXenocide** [comment]
We're in the process of changing how we deal with NuGet packages and this could be helpful for us. Honestly it would be very convenient if NuGetGallery was also a symbol server in the same way NuGet.org is; simply publish a symbol package and the symbols are handy for debug! Combine with Source Link and we have minimal configuration universal debugging. 

This is not on our short-term list though, just adding this to my wishlist and subscribing to updates :)

**Comment by loic-sharma** [comment]
@TheXenocide are you running your own private NuGet server? If so, some third-party NuGet servers also work as symbol servers like [Sleet ](https://github.com/emgarten/Sleet) or [BaGet](https://github.com/loic-sharma/BaGet). For a full list of third-party servers see this documentation: https://docs.microsoft.com/en-us/nuget/hosting-packages/overview

**Comment by TheXenocide** [comment]
@loic-sharma we currently use NuGet Gallery for our private nuget servers. We previously used the TeamCity integrated server but had significant performance issues with it which were causing other issues in our builds, so we separated from that and have just been using this project as our NuGet repository. We could consider an alternative, it just came to mind since NuGet.org has built-in symbol server support and I was thinking it would be nice to have it in our local version of the gallery as well. We're honestly a little ways out before we could consider employing any of this anyway (and would more particularly benefit from it once there's a TFVC implementation of Source Link), just thought I'd not that it would be nice in combination with this here.

**Comment by clairernovotny** [comment]
@TheXenocide you could also just include the PDB in your main nupkg and not worry about symbol servers at all. Add this to your props/csproj file:

https://github.com/clairernovotny/DeterministicBuilds/blob/master/Directory.Build.props#L10
